### ISSUE TITLE
Add rationale text to introduction

### ISSUE THREAD

**Issue by dthaler** [core-wg/coap-tcp-tls#50]
Currently section 1 motivates why TCP as a transport is needed, but doesn't motivate why using COAP (as opposed to HTTP/2) over TCP is desirable.  Such motivation should be added, since there's a document proposed in httpbis (draft-montenegro-httpbis-h2ot) to do an HTTP/2 profile for IoT and hence seems to target the same use cases.   So additional text should be added explaining the rationale for using coap even when going over tcp.

**Comment by cabo** [comment]
Fundamentally, the onus should be on the other draft to provide its rationale.  We certainly can expand the rationale to cover more about the advantages of CoAP over HTTP/2 for IoT applications.  I just think it might be a bit out of scope for the present draft, which is about CoAP over TCP, not about HTTP/2 over TCP or CoAP vs. HTTP/2.

**Comment by dthaler** [comment]
I completely disagree with cabo.   This is not about draft vs draft.  This is about each draft independently providing motivation for why it's a good solution to the problem.  This should not be hard.

**Comment by cabo** [comment]
But CoAP-over-TCP is quite evidently a good solution for the problem of running CoAP over TCP, TLS, and WebSockets.  I thought your observation was about the need for motivating the problem.  I don't want any single draft for some element of Technology A to turn into a diatribe why Technology A is better than Technology B.  (This, by the way, is exactly the problem with the above referenced draft, which adds to the insult by, instead of spending the pages motivating their Technology A, prefers spouting uninformed musings about Technology B.)

**Comment by cabo** [comment]
(Adding a sentence or two that further clarifies that the present draft is not intended to position CoAP as a replacement for the other REST protocols wouldn't hurt, though.)

**Comment by brianraymor** [comment]
Dave's request is perfectly reasonable. There's no reason for us to be defensive. With the proliferation of IoT protocols, there is confusion about what the IETF is recommending for IoT. In the [Internet of Things: Standards and Guidance from the IETF](https://www.internetsociety.org/publications/ietf-journal-april-2016/internet-things-standards-and-guidance-ietf) that Carsten co-authored, CoAP (+UDP) is described as **_a radically simplified UDP-based analog to HTTP_** - which would make CoAP (+TCP) a **_radically simplified TCP-based analog to HTTP_**.

In the same article, the reader learns that the  _HTTPbis WG recently [2015] finalized the specification for the HTTP/2 protocol that is **more suitable for IoT scenarios** than earlier versions of HTTP_.  If the reader does not have existing investments in CoAP, should HTTP/2 be preferred for TCP scenarios?

**Comment by dthaler** [comment]
An example of the sort of thing I'm asking for might be an argument that if a device already does COAP for proximal communication, but does not already do HTTP for any other reason, then a constrained node might benefit from coap over tcp due to code space constraints.   (Such an argument of course assumes there isn't significant overlap in code between COAP-over-TCP and something-else-over-TCP.)

**Comment by RobbySimpson** [comment]
@cabo states:
(This, by the way, is exactly the problem with the above referenced draft, which adds to the insult by, instead of spending the pages motivating their Technology A, prefers spouting uninformed musings about Technology B.)

I realize this is a bit off-topic for this issue, but considering this is a public forum and your statements are rather harsh, I'd love to see you expand on what you believe to be "uninformed," insulting, etc. about the referenced draft, preferably on-list.  The referenced draft attempts to give an overview of the plethora of IETF and other IoT standards and then goes on to discuss motivations and optimizations to HTTP/2.  If this is not the case, we'd love to know why.  Thanks!

**Comment by cabo** [comment]
In PR #51, I propose put in an ad for HTTP/2 (and XMPP, probably more important at this time) as other protocols that can be run over TCP/TLS.

**Comment by brianraymor** [comment]
@cabo - I've closed #51. We should focus on the motivation and rationale for CoAP+TCP rather than simply _advertising_ other protocols.

**Comment by cabo** [comment]
I don't get it.  The motivation and rationale for CoAP+TCP is there.  Any text that motivates selecting CoAP+TCP instead of HTTP/2 or XMPP requires mentioning these other protocols.  If we don't want to mention them, we are done.

**Comment by brianraymor** [comment]
The rationale for CoAP+TCP in reference to CoAP+UDP is clearly understood.

@dthaler is not requesting equal time to advertise HTTP or XMPP. He wants to understand why CoAP+TCP might be preferred over HTTP/2 in IoT scenarios. He even offered a sample rationale in one of his comments. Certainly companies with LWM2M investments have demonstrated a preference for shipping CoAP+TCP early because it has less impact on their deployments than transitioning to HTTP/2. 

There is an open source developer community supporting CoAP. There is not a similar community for HTTP/2 for IoT (yet).

**Comment by cabo** [comment]
There is a lot of rationale in RFC 7252.  Why do we need this here?  Why re HTTP/2 but not re XMPP? This is simply not rational.

**Comment by hannestschofenig** [comment]
I can give you the motivation why we are interested in CoAP over TLS /TCP. We have an existing implementation of LWM2M, which uses CoAP. We spent a lot of time getting that implementation rock-solid. Some enterprise deployments, which happen to have interesting firewall policies, do not allow us to use UDP. Hence, we were interested to add a TCP-based transport to CoAP. Making this enhancement turns out to be reasonably simple. 

As Carsten explained, CoAP has found its way into various SDOs and that's what people now implement. CoAP is sort-of understood and there are lots of implementations available. Switching to something else takes time and resources. There are also much fewer implementations of HTTP/2 for embedded devices available. 

Maybe in a few years we will be using HTTP/2 instead of CoAP over TCP/TLS but today we just want to get our existing deployments document and standardized. I am sure that people will take a closer look at the different options in more detail in the future. 

As soon as the CoAP over TCP / TLS spec is finished we will update our implementation.

**Comment by brianraymor** [comment]
How about adding the following to the **Introduction**:

... 
Some CoAP deployments need to integrate well with existing enterprise
infrastructures, where UDP-based protocols may not be well-received or may
even be blocked by firewalls. Middleboxes that are unaware of CoAP usage for
IoT can make the use of UDP brittle, resulting in lost or malformed packets.
...
_Emerging standards such as LWM2M [add reference] currently use CoAP over UDP as a transport
and require support for CoAP over TCP to address the issues above and to protect investments
in existing CoAP implementations and deployments. Although HTTP/2 could also potentially address these requirements, there would be additional costs and delays introduced by such a transition. There are also fewer HTTP/2 implementations available for constrained devices in comparison to CoAP._
